design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,299 @@
+
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
+	* alternatively, if currently running, take VM offline immediately and update machine type
+	* if not currently running, update machine type immediately
+* Allow user to specify certain VMs for updating machine type automation (e.g. by namespace)
+
+## Non Goals
+While these additions are beneficial to cluster-admins who wish to use the
+mass machine type transition (MMTT), these will not be included in the initial implementation:
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Subcommands that allow the user to monitor and manage the status of the MMTT job and affected VMs
+
+These will eventually be implemented in follow ups once the initial design is implemented.
+
+Additionally, this design proposal does not cover the method in which users/cluster admins
+will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all
+VMs with unsupported machine types without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace, label-selector).
+As a cluster-admin, I want to be in control of the behavior of running VMs being
+updated; I can determine whether I want to manually restart running VMs being
+updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command `virtctl update machine-types` that will allow
+the user to automatically update the machine type of any VMs with a machine type
+that is no longer supported. This command invokes a k8s `Job` that iterates
+through all VMs within a specified namespace (or all namespaces if none is
+specified), determines if the machine type is no longer supported, and updates
+it to the latest supported version. If the Job fails or is killed, it will be
+restarted with `restartPolicy=OnFailure` in the Job spec. The `restart-vm-required`
+label may then remain on a VM that has already been shut down, so the Job will
+first check VMs with the label and confirm if the machine types have been updated.
+If they have, then the label will be removed, but if not, the label will remain
+and the VM will continue to be tracked until it is restarted and its machine
+type is updated. Both the minimum supported machine type and the target machine
+type will be determined (retrieved) and configured internally when the command is
+invoked, and these values will automatically be updated as future versions are
+released and more machine types are deprecated upon these releases.
```

Comment:
How will we determine the correct machine type?

Reply:
Currently this is being manually handled and will need some more discussion before being fully implemented in a way that retrieves the correct machine type dynamically; this will be added in a follow-up implementation phase rather than the initial implementation. However, we are currently discussing how to expand the scope of this to not be vendor specific to RHEL/CentOS Stream, so it might end up being implemented anyway while solving this issue.

Reply:
Can we address this in the design?

Reply:
I have updated the design proposal with the proposed solution for determining the machine types. Please take a look at Itamar's comment on the draft PR for more context. [kubevirt/kubevirt#10222(comment)](https://github.com/kubevirt/kubevirt/pull/10222#issuecomment-1689491133)